runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -0,0 +1,1531 @@
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	j9object_t ptypes = J9VMJAVALANGINVOKEMETHODTYPE_PTYPES(currentThread, typeObject);
+	U_32 numArgs = J9INDEXABLEOBJECT_SIZE(currentThread, ptypes);
+
+	char* methodDescriptor = NULL;
+	char* cursor = NULL;
+	char* rSignature = NULL;
+
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	char** signatures = (char**)j9mem_allocate_memory((numArgs) * sizeof(char*), OMRMEM_CATEGORY_VM);
+	if (NULL == signatures) {
+		vm->internalVMFunctions->setNativeOutOfMemoryError(currentThread, 0, 0);
+		goto done;
+	}
+
+	*signatureLength = 2; /* space for '(', ')' */
+	for (U_32 i = 0; i < numArgs; i++) {
+		j9object_t pObject = J9JAVAARRAYOFOBJECT_LOAD(currentThread, ptypes, i);
+		J9Class *pclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, pObject);
+		signatures[i] = sigForPrimitiveOrVoid(currentThread, pclass);
+		if (VM_VMHelpers::exceptionPending(currentThread)) {
+			goto done;
+		} else if (NULL == signatures[i]) {
+			signatures[i] = getClassSignature(currentThread, pclass);
+			if (VM_VMHelpers::exceptionPending(currentThread)) {
+				goto done;
+			}
+		}
+
+		*signatureLength += strlen(signatures[i]);
+	}
+
+	{
+		/* Return type */
+		j9object_t rtype = J9VMJAVALANGINVOKEMETHODTYPE_RTYPE(currentThread, typeObject);
+		J9Class *rclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, rtype);
+		rSignature = sigForPrimitiveOrVoid(currentThread, rclass);
+		if (VM_VMHelpers::exceptionPending(currentThread)) {
+			goto done;
+		} else if (rSignature == NULL) {
+			rSignature = getClassSignature(currentThread, rclass);
+			if (VM_VMHelpers::exceptionPending(currentThread)) {
+				goto done;
+			}
+		}
+	}
+
+	*signatureLength += strlen(rSignature);
+
+	methodDescriptor = (char*)j9mem_allocate_memory(*signatureLength+1, OMRMEM_CATEGORY_VM);
+	if (NULL == methodDescriptor) {
+		vm->internalVMFunctions->setNativeOutOfMemoryError(currentThread, 0, 0);
+		goto done;
+	}
+	cursor = methodDescriptor;
+	*cursor++ = '(';
+
+	/* Copy class signatures to descriptor string */
+	for (U_32 i = 0; i < numArgs; i++) {
+		U_32 len = strlen(signatures[i]);
+		strncpy(cursor, signatures[i], len);
+		cursor += len;
+	}
+
+	*cursor++ = ')';
+	/* Copy return type signature to descriptor string */
+	strcpy(cursor, rSignature);
+
+done:
+	if (NULL != rSignature) {
+		j9mem_free_memory(rSignature);
+	}
+	if (NULL != signatures) {
+		for (U_32 i = 0; i < numArgs; i++) {
+			if (NULL != signatures[i]) {
+				j9mem_free_memory(signatures[i]);
+			} else {
+				break;
+			}
+		}
+		j9mem_free_memory(signatures);
+	}
+	return methodDescriptor;
+}
+
+j9object_t
+resolveRefToObject(J9VMThread *currentThread, J9ConstantPool *ramConstantPool, U_16 cpIndex, bool resolve)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	const J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+
+	j9object_t result = NULL;
+
+	J9RAMSingleSlotConstantRef *ramCP = (J9RAMSingleSlotConstantRef*)ramConstantPool + cpIndex;
```

Comment:
It is better to get constants {int, float, double, long} from the ROM cp as they may be in the RAM cp.  When possible, we put constant data to the end of the ROM CP and have smaller RAM CPs.  Always fetch that kind of data from the ROM cp entries - see `getSingleSlotConstant` or `getDouleSlotConstant` from sun_reflect_constantpool.c